persistent/Database/Persist/Sql/Orphan/PersistUnique.hs

```diff
@@ -18,43 +21,75 @@ import qualified Data.Conduit.List as CL
-                                           
-                                    go'' n Assign = n <> "=?"
-                                    go'' n Add = T.concat [n, "=", n, "+?"]
-                                    go'' n Subtract = T.concat [n, "=", n, "-?"]
-                                    go'' n Multiply = T.concat [n, "=", n, "*?"]
-                                    go'' n Divide = T.concat [n, "=", n, "/?"]
-                                    go'' _ (BackendSpecificUpdate up) = error $ T.unpack $ "BackendSpecificUpdate" `mappend` up `mappend` "not supported"
-                                              
-                                    go' (x, pu) = go'' (connEscapeName conn x) pu
-                                    go x = (fieldDB $ updateFieldDef x, updateUpdate x)
-
-                                x <- rawSql sql vals
-                                return $ head x
-        Nothing -> defaultUpsert record updates
-        where
-          t = entityDef $ Just record
-          unqs uniqueKey = concat $ map (persistUniqueToValues) [uniqueKey]
-
+        conn <- ask
+        uniqueKey <- onlyUnique record
+        case connUpsertSql conn of
+            Just upsertSql ->
+                case updates of
+                    [] -> defaultUpsert record updates
+                    xs -> do
+                        let upds = T.intercalate "," $ map (go' . go) updates
+                            sql = upsertSql t upds
+                            vals =
+                                (map toPersistValue $ toPersistFields record) ++
+                                (map updatePersistValue updates) ++
+                                (unqs uniqueKey)
+                            go'' n Assign =
+                                escape (entityDB t) <> "." <> n <> "=?"
+                            go'' n Add =
+                                T.concat
+                                    [ n
+                                    , "="
+                                    , escape (entityDB t) <> "."
+                                    , n
+                                    , "+?"]
+                            go'' n Subtract =
+                                T.concat
+                                    [ n
+                                    , "="
+                                    , escape (entityDB t) <> "."
+                                    , n
+                                    , "-?"]
+                            go'' n Multiply =
+                                T.concat
+                                    [ n
+                                    , "="
+                                    , escape (entityDB t) <> "."
+                                    , n
+                                    , "*?"]
+                            go'' n Divide = T.concat [n, "=", n, "/?"]
```

Comment:
Good catch, I will fix this and probably make it generic as you have suggested. The previous code, I think was adapted from somewhere else in the same codebase.